### PR TITLE
💡 Make validation method post-init & parameter-less

### PR DESCRIPTION
**Pull request by atifaziz**
This PR demonstrates the idea discussed in [my comment](https://github.com/amis92/RecordGenerator/issues/54#issuecomment-519404007) to #54.

One side benefit of a parameter-less `Validate` method is that the syntax trees for the declaration and invocation can be shared singletons.

One new downside I discovered while taking the idea for a spin is that argument names are lost and so you have to quote and sync them manually instead of being able to rely on `nameof`. 😢 

If you like & agree with the idea, then feel free to merge the PR. Otherwise I consider it to be a demonstration/spike PR that can be closed if we conclude that this doesn't cut it.

### CODE REVIEW

File: test/Amadevus.RecordGenerator.TestsBase/FeaturesTestType.cs [hunk 1]
```diff
@@ -11,9 +11,9 @@ public partial class FeatureConstructor
             // and we won't see it in reflection anyway.
             // But since we have to implement it, it's already testing
             // that the generated partial declaration exists during compilation.
-            static partial void Validate(ref string name)
+            partial void Validate()
             {
-                if (name is null) throw new System.ArgumentNullException(nameof(name));
+                if (Name is null) throw new System.ArgumentNullException("name");
```
Comment (amis92) [inline test/Amadevus.RecordGenerator.TestsBase/FeaturesTestType.cs]: I'd keep this as `nameof(Name)`. Different casing is not a huge issue here IMO, but more importantly any refactoring won't "break" this.

File: test/Amadevus.RecordGenerator.TestsBase/ValidatingRecord.cs [hunk 2]
```diff
@@ -14,9 +14,9 @@ public partial class ValidatingRecord
         /// </summary>
         public string Switch { get; }
 
-        static partial void Validate(ref string name, ref string @switch)
+        partial void Validate()
         {
-            if (name is null) throw new ArgumentNullException(nameof(name));
+            if (Name is null) throw new ArgumentNullException("name");
```
Comment (amis92) [inline test/Amadevus.RecordGenerator.TestsBase/ValidatingRecord.cs]: I'd keep this as `nameof(Name)`. Different casing is not a huge issue here IMO, but more importantly any refactoring won't "break" this.

**Comment by amis92** [comment]
In the end, I think this is the most reasonable approach to validation. After discussions in #54 I think it's the simplest of all possible approaches. It also lends itself quite well for future enhancements in other areas.

It does keep the cost of running validation on smallest changes, but the cost of avoiding it seems to be a little overblown (https://github.com/amis92/RecordGenerator/issues/54#issuecomment-519417538)

**Comment by atifaziz** [comment]
> Just a nitpick to keep using `nameof` instead of literal string.

You mean this is just for this project, right? Any other project would have to hard-wire them?

I don't think we're completely out of the woods with this. Have a look at d0eff410d12e2b8797bc7d5e856194873f658f6a. The new value parameter for all `With*` members is named `value` and seeing an `ArgumentNullException` where its `ParamName` reads something else can be confusing. The demonstration in d0eff410d12e2b8797bc7d5e856194873f658f6a deliberately shows a breaking test.

In the end, I am undecided if it's a real issue or not because it's okay for methods to delegate to deeper methods that may validate and which may in turn have different parameter names.

File: test/Amadevus.RecordGenerator.Test/RecordMutatorTests.cs [hunk 3]
```diff
@@ -27,6 +28,13 @@ public void With_SimpleProperty_CreatesModifiedInstance()
             Assert.Equal(newName, modifiedItem.Name);
         }
 
+        [Fact]
+        public void Ctor_InvokesValidate_Passing()
+        {
+            var item = CreateItem();
+            Assert.Throws<ArgumentNullException>("value", () => item.WithName(null));
```
Comment (amis92) [inline test/Amadevus.RecordGenerator.Test/RecordMutatorTests.cs]: I'm not sure what you're trying to demo. Argument name passed to `ArgumentNullException` should reflect argument name at the place of check being done. The check is done in constructor via delegation to partial `Validate` method.

`Validate` body is written by the user, doing whatever he'll want to. And in this case I'd say that calling out Property name for the exception is as correct as it gets, because the ctor right now is just a bridge between caller and property setter. Ideally, this'd be done in the object initializer, but we can't use it 'cause we're immutable.

So to sum up, let's use nameof(Property) for our tests. In user guidelines that's also what we can advertise. And then, in the end, it's up to a developer to write the exception code.

**Comment by amis92** [comment]
> In the end, I am undecided if it's a real issue or not because it's okay for methods to delegate to deeper methods that may validate and which may in turn have different parameter names.

That's exactly what I'd focus on. We're delegating checks deeper because of circumstances. The checks should use property name as that's what they're *actually* verifying, as well as because there are multiple scenarios that can throw this exception:

```csharp
[Record]
public partial class MyRecord
{
  public string Name { get; }
  partial void Validate() {
    if (Name is null) throw new ArgumentNullException("???");
  }
}

[Fact]
public void Tests()
{
  var record = new MyRecord(null); // throws, parameter name is 'name'
  record.WithName(null); // throws, parameter name is 'value'
  new MyRecord.Builder { Name = null }.ToImmutable(); // throws, no actual parameter per se
}
```

File: test/Amadevus.RecordGenerator.Test/RecordMutatorTests.cs [hunk 4]
```diff
@@ -32,7 +32,14 @@ public void With_SimpleProperty_CreatesModifiedInstance()
...
             var item = CreateItem();
-            Assert.Throws<ArgumentNullException>("value", () => item.WithName(null));
+            //
+            // NOTE! We test that validation is indeed invoked but don't
+            // test the parameter name since it comes from the validation
+            // method that assumes the name of the corresponding parameter on
+            // the constructor ("name") as opposed to WithName (the parameter
+            // of all with-methods is always named "value").
+            //
+            Assert.Throws<ArgumentNullException>(() => item.WithName(null));
```
Comment (amis92) [inline test/Amadevus.RecordGenerator.Test/RecordMutatorTests.cs]: Let's assume it's `nameof(Item.Name)`.

**Comment by atifaziz** [comment]
> So to sum up, let's use `nameof(Property)` for our tests. In user guidelines that's also what we can advertise. And then, in the end, it's up to a developer to write the exception code.

Have a look at 4076ebde486b317355b0a1e2b2e9f935a53a4050, where validation becomes this:

```c#
        partial void Validate() =>
            Validate(Name);

        static void Validate(string name)
        {
            if (name == null) throw new ArgumentNullException(nameof(name));
        }
```

So you delegate to another method to whom you pass only the fields needing validation. The additional helper validation has the effect of then also capturing the argument names for exception via `nameof`. What do you think? I feel it strikes a nice sweet spot and the test represents guidance too. 😄

**Comment by atifaziz** [comment]
I think rather than being prescriptive with a name like `Validate`, we should just call it `OnConstructed` or `OnInitialized`, which is just what it is. You can then do what you want in there, like validation. It also makes the following read better:

```c#
partial void OnInitialized() =>
    Validate(Name);

static void Validate(string name)
{
    if (name == null) throw new ArgumentNullException(nameof(name));
}
```

How do you feel about this?

File: test/Amadevus.RecordGenerator.Test/RecordMutatorTests.cs (hunk 3 above)
Comment (atifaziz) [reply test/Amadevus.RecordGenerator.Test/RecordMutatorTests.cs]: See 4076ebde486b317355b0a1e2b2e9f935a53a4050 for a compromise.

**Comment by amis92** [comment]
I like the change of the name to `OnConstructed`.

Other than that, I'd call the rest as unimportant details. But for a guidance, I think it becomes over-engineered. If the validation is more complex, sure. But for the common case, I'd keep the exception throwing within the `OnConstructed` method.

As for the `nameof` usage, I think that using `nameof(Property)` is a good balance between throwing `ArgumentNullException` without argument name, and selecting arbitrary name (as is the case with the additional `ValidateName(string)`).

Property name is the thing that matters. Imagine seeing the stack trace. If the exception is thrown from within OnConstructed method with no parameters, it's a good guess that it was validating it's calling method, in this case constructor. And the constructor itself can be called by any With- method, ToImmutable, or some factory method user wrote.

---

For tests, let's use simple `OnConstructed()` with a body that throws `ArgumentNullException(nameof(Property))`.

For future guidance, do the same. Maybe suggest using custom `ValidateProperty` method to add a more meaningful stack trace entry.

File: test/Amadevus.RecordGenerator.Test/RecordMutatorTests.cs (hunk 3 above)
Comment (amis92) [reply test/Amadevus.RecordGenerator.Test/RecordMutatorTests.cs]: I'll stay with the `nameof(Property)`. It's more maintainable for our tests, and doesn't strike me as misleading anyway.

**Comment by atifaziz** [comment]
> I like the change of the name to `OnConstructed`.

Done with f0e212246824379babc725a5727fe166a373ad19.

File: test/Amadevus.RecordGenerator.Test/RecordMutatorTests.cs (hunk 4 above)
Comment (atifaziz) [reply test/Amadevus.RecordGenerator.Test/RecordMutatorTests.cs]: Done with 1ff6ba45e45245343dcc73510986db2dd5f976ba.

File: test/Amadevus.RecordGenerator.Test/RecordMutatorTests.cs (hunk 3 above)
Comment (atifaziz) [reply test/Amadevus.RecordGenerator.Test/RecordMutatorTests.cs]: Done with 1ff6ba45e45245343dcc73510986db2dd5f976ba.
